torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1398,7 +1398,7 @@ void InitXlaModuleBindings(py::module m) {
     if (xtensor->CurrentIrValue()) {
       device_data_node = DeviceData::Cast(xtensor->CurrentIrValue().node.get());
       if (!device_data_node) {
-        xtensor->SetShardingSpec(*new_sharding_spec);
+        tensor_methods::custom_sharding_(xtensor, new_sharding_spec);
```

Comment:
I vaguely remember that we discussed about  for IR we want to do custom sharding and for input/output we want to directly injected the sharding spec to the HLO. Is that still the case?

Reply:
Yea, output means the ROOT node which should not be affected by mark_sharding.

Reply:
hmm then how does output get the sharding? Through in-place update and sharding spec being inherited to the output tensor? I remember there are cases where we want to manually mark sharding on output.

Reply:
There is a logic to extract the sharding spec from the output and then assign it to the output tensors: https://github.com/pytorch/xla/blob/master/torch_xla/csrc/xla_sharding_util.cpp#L629

This logic works no matter whether the ROOT node has sharding spec or not, i.e., whether or not the output sharding propagation is enabled.

Reply:
so the idea here is that only `mark_sharding` and functionization should use `custom_sharding`? I saw there are a couple other places like in `clone` still use `setShardingSpec`.

For data through data loader I think it is attach sharding to HLO directly
For model weights, we will use `mark_sharding` so `custom_sharding`
For output it depends on whether `mark_sharding` is being called on the output explictly
For Clone it will be directly attach sharding to the IR cloned.

Is this the right understanding?

Reply:
The custom_sharding op should only be applied to IR node which means not inputs or outputs.

Data from data loader and weights are inputs so they will use setShardingSpec directly.
Output with output sharding propagation, we don't mark anything. Output means ROOT node. Without propagation, we will do setShardingSpec directly.

For Clone, I have no ideas... Can you point me to the code?

Reply:
[I am glad we get to meet a couple times during TGIT, that was fun. ](https://github.com/pytorch/xla/blob/4f378e694a68192fd634a070a3485c73afab1c71/torch_xla/csrc/tensor_methods.cpp#L906-L912)

Reply:
Got it, I think it should be custom-call. In fact, all SetShardingSpec except inputs and outputs should be replaced by custom-calls. But I want to do it as a follow up after we were able to reproduce the LLaMA2 results on the RC wheels.

Reply:
File #5495.

Reply:
For reference, clone, copy -- and even functionalization related path should be annotated with custom call.